<a href="https://colab.research.google.com/github/AstridSerruto/Projects/blob/master/Predicting/Predicting_Football_matches.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predicting football matches for the English League

In [31]:
# Importing pandas
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

In [32]:
#Obtaining data

from google.colab import drive
drive.mount('/content/drive')
	
matches = pd.read_csv('/content/drive/My Drive/Datasets/matches.csv', index_col=0)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [33]:
#Checking table contents

matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [34]:
matches.shape

(1389, 27)

In [35]:
# Checking basic info of the dataset

missing_stats = []

for col in matches.columns:
    missing_stats.append((col, matches[col].nunique(), matches[col].isnull().sum() / matches.shape[0], matches[col].value_counts(normalize=True, dropna=False).values[0], matches[col].dtype))
    
stats_df = pd.DataFrame(missing_stats, columns=['feature', 'unique_values','percent_missing', 'percent_biggest_cat', 'type'])
stats_df

feature  unique_values  percent_missing  percent_biggest_cat     type
0           date            244          0.00000             0.014399   object
1           time             18          0.00000             0.205904   object
2           comp              1          0.00000             1.000000   object
3          round             38          0.00000             0.028078   object
4            day              7          0.00000             0.396688   object
5          venue              2          0.00000             0.500360   object
6         result              3          0.00000             0.394528   object
7             gf              9          0.00000             0.326854  float64
8             ga              9          0.00000             0.326134  float64
9       opponent             23          0.00000             0.051116   object
10            xg             44          0.00000             0.063355  float64
11           xga             45          0.00000             0.063355  float64
12          poss             64          0.00000             0.035277  float64
13    attendance            334          0.50108             0.501080  float64
14       captain             91          0.00000             0.050396   object
15     formation             16          0.00000             0.247660   object
16       referee             23          0.00000             0.071994   object
17  match report              1          0.00000             1.000000   object
18         notes              0          1.00000             1.000000  float64
19            sh             32          0.00000             0.087833  float64
20           sot             16          0.00000             0.187185  float64
21          dist            160          0.00072             0.018719  float64
22            fk              5          0.00000             0.629230  float64
23            pk              4          0.00000             0.889129  float64
24         pkatt              4          0.00000             0.863931  float64
25        season              2          0.00000             0.547156    int64
26          team             23          0.00000             0.051836   object

In [36]:
#How many matches for each team

matches['team'].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

In [37]:
#Why did Liverpool have less matches?

matches[matches['team'] =='Liverpool']

#We are missing season.

date   time            comp         round  day venue result   gf  \
1   2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13  2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14  2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16  2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18  2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22  2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23  2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24  2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25  2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27  2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28  2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30  2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31  2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32  2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33  2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34  2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36  2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37  2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38  2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41  2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42  2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44  2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46  2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47  2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk  \
1   3.0     Leeds United  ...  Match Report    NaN  20.0   4.0  17.0  0.0   
2   0.0          Chelsea  ...  Match Report    NaN  17.0   5.0  17.7  1.0   
4   1.0          Arsenal  ...  Match Report    NaN  21.0   9.0  16.8  0.0   
6   7.0      Aston Villa  ...  Match Report    NaN  14.0   8.0  15.8  1.0   
7   2.0          Everton  ...  Match Report    NaN  22.0   8.0  15.0  1.0   
9   1.0    Sheffield Utd  ...  Match Report    NaN  17.0   5.0  18.2  1.0   
11  1.0         West Ham  ...  Match Report    NaN   8.0   2.0  18.6  1.0   
13  1.0  Manchester City  ...  Match Report    NaN   9.0   2.0  21.5  0.0   
14  0.0   Leicester City  ...  Match Report    NaN  24.0  12.0  11.9  0.0   
16  1.0         Brighton  ...  Match Report    NaN   6.0   2.0  20.9  0.0   
18  0.0           Wolves  ...  Match Report    NaN  11.0   6.0  16.6  1.0   
20  1.0           Fulham  ...  Match Report

In [38]:
matches['round'].value_counts()

Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: round, dtype: int64

Some indicate that we have fewer matches than 39, due to the fact that we web scrapped the data while the season was still going.

In [39]:
#Change date to datetime type

matches['date'] = pd.to_datetime(matches['date'])


Cleaning the data

In [40]:
#Convert to categorical data and then to int

matches['venue_code'] = matches['venue'].astype('category').cat.codes

In [41]:
#Create category for each opponent and then convert to int

matches['opp_code'] = matches['opponent'].astype('category').cat.codes

In [42]:
#Replace ; and minutes with numbers, convert to int

matches['hour'] = matches['time'].str.replace(':.+', "", regex=True).astype('int')

In [43]:
#Obtain number for each day of the week

matches['day_code'] = matches['date'].dt.dayofweek

In [64]:
#Set target for algorithmn, 0= L or D, 1= W

matches["target"] = (matches["result"] == "W").astype("int")

In [65]:
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   fk   pk  pkatt  season              team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022   Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022   Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022   Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022   Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022   Manchester City   
..  ...             ...  ...  ...  ...    ...     ...               ...   
38  4.0       Tottenham  ...  0.0  0.0    0.0    2021  Sheffield United   
39  2.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Sheffield United   
40  0.0         Everton  ...  0.0  0.0    0.0    2021  Sheffield United   
41  1.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Sheffield United   
42  0.0         Burnley  ...  0.0  0.0    0.0    2021  Sheffield United   

   venue_code opp_code hour  day_code  target  
1           0       18   16         6       0  
2           1       15   15         5       1  
3           1        0   12         5       1  
4           0       10   15         5       1  
6           1       17   15         5       0  
..        ...      ...  ...       ...     ...  
38          0       18   19         6       0  
39          1        6   15         5       0  
40          0        7   19         6       1  
41          0       14   18         2       0  
42          1        4   16         6       1  

[1389 rows x 32 columns]

## Create initial Machine Learning model

In [46]:
#Import classifier

from sklearn.ensemble import RandomForestClassifier

In [47]:
#Initialize the class and pass parameters

rf = RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)

In [66]:
#Split the data into train and test

train = matches[matches['date'] < '2022-01-01']

In [67]:
test = matches[matches['date'] > '2022-01-01']

In [49]:
#Create predictors

predictors = ['venue_code', 'opp_code', 'hour', 'day_code']

In [50]:
#Fit the model

rf.fit(train[predictors], train['target'])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [51]:
#Generate predictions

preds = rf.predict(test[predictors])

In [52]:
#Import metrics

from sklearn.metrics import accuracy_score

In [53]:
#Pass actual and predictions

acc = accuracy_score(test['target'], preds)

In [54]:
acc

0.6123188405797102

In [55]:
#Create DataFrame to combine actual and predicted values

combined = pd.DataFrame(dict(actual=test['target'], prediction=preds))

In [56]:
#Create crosstap

pd.crosstab(index=combined['actual'], columns=combined['prediction'])

prediction    0   1
actual             
0           141  31
1            76  28

In [68]:
#We need to revise our metric to improve our prediction
#Import precision score

from sklearn.metrics import precision_score

In [69]:
precision_score(test['target'], preds)

0.4745762711864407

As we can see, accuracy(61.23%) and precision(47.45%) are not very good with this model. We will make improvements with rolling averages.

In [71]:
#Group matches by team

grouped_matches = matches.groupby('team')

In [72]:
#Now we can see each team as a table

group = grouped_matches.get_group('Manchester City')
group

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
54 2021-05-01  12:30  Premier League  Matchweek 34  Sat  Away      W  2.0   
56 2021-05-08  17:30  Premier League  Matchweek 35  Sat  Home      L  1.0   
57 2021-05-14  20:00  Premier League  Matchweek 36  Fri  Away      W  4.0   
58 2021-05-18  19:00  Premier League  Matchweek 37  Tue  Away      L  2.0   
59 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  5.0   

     ga        opponent  ...   fk   pk  pkatt  season             team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022  Manchester City   
..  ...             ...  ...  ...  ...    ...     ...              ...   
54  0.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Manchester City   
56  2.0         Chelsea  ...  0.0  0.0    1.0    2021  Manchester City   
57  3.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Manchester City   
58  3.0        Brighton  ...  1.0  0.0    0.0    2021  Manchester City   
59  0.0         Everton  ...  0.0  0.0    0.0    2021  Manchester City   

   venue_code opp_code hour  day_code  target  
1           0       18   16         6       0  
2           1       15   15         5       1  
3           1        0   12         5       1  
4           0       10   15         5       1  
6           1       17   15         5       0  
..        ...      ...  ...       ...     ...  
54          0        6   12         5       1  
56          1        5   17         5       0  
57          0       14   20         4       1  
58          0        3   19         1       0  
59          1        7   16         6       1  

[71 rows x 32 columns]

In [75]:
#Assign new rolling averages to columns

def rolling_averages(group, cols, new_cols):
  group = group.sort_values('date') #Sort groups by date
  rolling_stats = group[cols].rolling(3, closed='left').mean() #Compute rolling averages for columns
  group[new_cols] = rolling_stats #Assign stats to original DataFrame
  group = group.dropna(subset=new_cols) #Drop missing values
  return group

In [76]:
#Define the columns

cols = ['gf', 'ga', 'sh', 'sot', 'dist', 'fk', 'pk', 'pkatt']
new_cols = [f'{c}_rolling' for c in cols]
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [77]:
#Call averages for a single group

rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  day_code  target  gf_rolling  ga_rolling  \
5   0.0         Arsenal  ...         5       1    2.000000    2.333333   
7   1.0        West Ham  ...         5       0    1.333333    2.000000   
9   0.0   Sheffield Utd  ...         5       1    1.000000    0.666667   
11  1.0       Liverpool  ...         6       0    1.000000    0.333333   
12  2.0       Tottenham  ...         5       0    1.000000    0.666667   
..  ...             ...  ...       ...     ...         ...         ...   
42  0.0  Crystal Palace  ...         0       0    2.333333    1.333333   
44  0.0         Burnley  ...         5       1    1.666667    0.333333   
46  2.0       Liverpool  ...         6       0    2.000000    0.333333   
49  0.0        Brighton  ...         2       1    1.333333    0.666667   
50  1.0         Watford  ...         5       1    2.333333    0.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5   17.333333    4.666667    18.900000   1.333333    0.333333       0.333333  
7   17.333333    3.666667    17.733333   0.666667    0.000000       0.000000  
9   16.666667    4.333333    18.233333   0.666667    0.000000       0.000000  
11  14.333333    6.666667    18.466667   1.000000    0.000000       0.000000  
12  12.000000    5.666667    19.366667   1.000000    0.000000       0.333333  
..        ...         ...          ...        ...         ...            ...  
42  19.000000    7.000000    15.366667   0.333333    0.333333       0.333333  
44  18.333333    7.333333    16.000000   0.333333    0.000000       0.000000  
46  20.000000    6.666667    16.133333   0.333333    0.000000       0.000000  
49  15.666667    4.666667    16.700000   0.333333    0.000000       0.000000  
50  15.333333    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

In [78]:
#Apply a function to each team

matches_rolling = matches.groupby('team').apply(lambda x: rolling_averages(x, cols, new_cols))
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling

In [79]:
#Drop the extra index level

matches_rolling = matches_rolling.droplevel('team')
matches_rolling

date   time            comp         round  day venue result   gf  \
6  2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
7  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
32 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
33 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
34 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
35 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
36 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
6   1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
7   1.0  Manchester City  ...         5       0    1.666667    1.666667   
9   1.0   Leicester City  ...         6       0    1.000000    1.666667   
11  0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
13  3.0      Aston Villa  ...         6       0    0.333333    0.666667   
..  ...              ...  ...       ...     ...         ...         ...   
32  0.0          Everton  ...         6       1    1.333333    1.000000   
33  3.0     Leeds United  ...         4       0    1.666667    0.666667   
34  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
35  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
36  1.0          Burnley  ...         6       0    1.333333    1.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6    7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
7    5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
9    7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
11   9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
13   9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
..        ...         ...          ...        ...         ...            ...  
32  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
33  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
34  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
35  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
36  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

In [80]:
#Assign values starting at 0
matches_rolling.index = range(matches_rolling.shape[0])
matches_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

       ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
1     1.0  Manchester City  ...         5       0    1.666667    1.666667   
2     1.0   Leicester City  ...         6       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
4     3.0      Aston Villa  ...         6       0    0.333333    0.666667   
...   ...              ...  ...       ...     ...         ...         ...   
1312  0.0          Everton  ...         6       1    1.333333    1.000000   
1313  3.0     Leeds United  ...         4       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
1316  1.0          Burnley  ...         6       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

## Retraining the model

In [81]:
def make_predictions(data, predictors):
  train = data[data['date'] < '2022-01-01']
  test = data[data['date'] > '2022-01-01']
  rf.fit(train[predictors], train['target'])
  preds =rf.predict(test[predictors])
  combined = pd.DataFrame(dict(actual=test['target'], predicted=preds), index=test.index)
  precision = precision_score(test['target'],preds)
  return combined, precision

In [82]:
#Call the function

combined, precision = make_predictions(matches_rolling, predictors + new_cols)
precision

0.625

Precision has improved from 47.45% to 62.5.

In [84]:
#Add information to combined

combined = combined.merge(matches_rolling[['date', 'team', 'opponent', 'result']], left_index=True, right_index=True)
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

In [85]:
#Normalize the data

class MissingDict(dict):
  __missing__ = lambda self, key: key
map_values = {
    'Brighton and Hove Albion': 'Brighton',
    'Manchester United': 'Manchester Utd',
    'Newcastle United': 'Newcastle Utd',
    'Tottenham Hotspur': 'Tottenham',
    'West Ham United': 'Est Ham',
    'Wolverhampton Wanderers': 'Wolves'
}
mapping = MissingDict(**map_values)

In [87]:
combined['new_team'] = combined['team'].map(mapping)

In [90]:
merged = combined.merge(combined, left_on=['date', 'new_team'], right_on=['date','opponent'])
merged

actual_x  predicted_x       date                   team_x     opponent_x  \
0           0            0 2022-01-23                  Arsenal        Burnley   
1           1            0 2022-02-10                  Arsenal         Wolves   
2           1            0 2022-02-19                  Arsenal      Brentford   
3           1            1 2022-02-24                  Arsenal         Wolves   
4           1            1 2022-03-06                  Arsenal        Watford   
..        ...          ...        ...                      ...            ...   
244         1            0 2022-03-13  Wolverhampton Wanderers        Everton   
245         0            0 2022-03-18  Wolverhampton Wanderers   Leeds United   
246         1            0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
247         0            0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
248         0            0 2022-04-24  Wolverhampton Wanderers        Burnley   

    result_x new_team_x  actual_y  predicted_y                   team_y  \
0          D    Arsenal         0            0                  Burnley   
1          W    Arsenal         0            0  Wolverhampton Wanderers   
2          W    Arsenal         0            0                Brentford   
3          W    Arsenal         0            0  Wolverhampton Wanderers   
4          W    Arsenal         0            0                  Watford   
..       ...        ...       ...          ...                      ...   
244        W     Wolves         0            0                  Everton   
245        L     Wolves         1            0             Leeds United   
246        W     Wolves         0            0              Aston Villa   
247        L     Wolves         1            0         Newcastle United   
248        L     Wolves         1            0                  Burnley   

    opponent_y result_y     new_team_y  
0      Arsenal        D        Burnley  
1      Arsenal        L         Wolves  
2      Arsenal        L      Brentford  
3      Arsenal        L         Wolves  
4      Arsenal        L        Watford  
..         ...      ...            ...  
244     Wolves        L        Everton  
245     Wolves        W   Leeds United  
246     Wolves        L    Aston Villa  
247     Wolves        W  Newcastle Utd  
248     Wolves        W        Burnley  

[249 rows x 13 columns]

In [91]:
#Look at rows were one team was predicted to win and the othe lose.

merged[(merged['predicted_x'] == 1) & (merged['predicted_y'] == 0)]['actual_x'].value_counts()

1    25
0    13
Name: actual_x, dtype: int64

In [92]:
27/40

0.675

Precision 67.5%